File: Day #1224/solution.go
// DFS with BST pruning: skip left if val<a, skip right if val>b. O(n) worst.
// O(n) time worst, O(h) space (recursion).
package main

import "fmt"

type Node struct {
	Val         int
	Left, Right *Node
}

func rangeSumBST(node *Node, a, b int) int {
	if node == nil {
		return 0
	}
	if node.Val < a {
		return rangeSumBST(node.Right, a, b)
	}
	if node.Val > b {
		return rangeSumBST(node.Left, a, b)
	}
	return node.Val + rangeSumBST(node.Left, a, b) + rangeSumBST(node.Right, a, b)
}

func main() {
	root := &Node{
		Val:   5,
		Left:  &Node{Val: 3, Left: &Node{Val: 2}, Right: &Node{Val: 4}},
		Right: &Node{Val: 8, Left: &Node{Val: 6}, Right: &Node{Val: 10}},
	}
	fmt.Println(rangeSumBST(root, 4, 9))
}

File: Day #1224/solution.py
# Day 1224: DFS with BST pruning: skip left if val<a, skip right if val>b. O(n) worst.
# O(n) time worst, O(h) space (recursion).
from typing import Optional


class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def range_sum_bst(node: Optional[Node], a: int, b: int) -> int:
    if not node:
        return 0
    if node.val < a:
        return range_sum_bst(node.right, a, b)
    if node.val > b:
        return range_sum_bst(node.left, a, b)
    return node.val + range_sum_bst(node.left, a, b) + range_sum_bst(node.right, a, b)


if __name__ == "__main__":
    root = Node(5, Node(3, Node(2), Node(4)), Node(8, Node(6), Node(10)))
    print(range_sum_bst(root, 4, 9))
